### PR TITLE
feat: Added support for multiple accounts

### PR DESCRIPTION
**Pull request by cdnninja**
This adds multiple account support.  Only the first car per account though.  This is modeled after dalhb's work.   Config is auto updated from the current build.  I have only tested this once though.  I plan to keep testing but wanted to get a PR in for review.

### CODE REVIEW

**Comment by cdnninja** [comment]
This will need squash and merge to ensure it doesn't show so many commits.

File: custom_components/kia_uvo/__init__.py [hunk 1]
```diff
@@ -71,15 +72,30 @@
 
 
 async def async_setup(hass: HomeAssistant, config_entry: ConfigEntry):
-    if DOMAIN not in hass.data:
-        hass.data[DOMAIN] = {}
+    hass.data.setdefault(DOMAIN, {})
+
+    def convert_call_to_vehicle(call) -> Vehicle:
+        vehicle_identifiers = list(hass.data[DOMAIN].keys())
+        if len(vehicle_identifiers) == 1:
```
Comment (fuatakgun) [inline custom_components/kia_uvo/__init__.py:79]: i would get rid of `if` condition here, `else` should work for all conditions.

File: custom_components/kia_uvo/__init__.py [hunk 2]
```diff
@@ -71,15 +72,30 @@
...
+            vehicle_identifier = vehicle_identifiers[0]
+        else:
+            vehicle_identifier = convert_device_id_to_vehicle_identifier(
+                call.data[ATTR_DEVICE_ID]
+            )
+
+        return hass.data[DOMAIN][vehicle_identifier][DATA_VEHICLE_INSTANCE]
+
+    def convert_device_id_to_vehicle_identifier(device_id: str) -> str:
```
Comment (fuatakgun) [inline custom_components/kia_uvo/__init__.py:88]: would rename `get_vehicle_identifier_from_device_id`

File: custom_components/kia_uvo/__init__.py [hunk 3]
```diff
@@ -244,12 +262,29 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry):
...
 
     return unload_ok
+
+
+async def async_migrate_entry(hass, config_entry: ConfigEntry):
+    if config_entry.version == 1:
+
+        vehicle_id = config_entry.data["stored_credentials"]["vehicle_id"]
+        new_data = config_entry.data.copy()
+        new_data[CONF_VEHICLE_IDENTIFIER] = vehicle_id
```
Comment (fuatakgun) [inline custom_components/kia_uvo/__init__.py:284]: looking here, vehicle_id should be equal to device_id. So, are these equal?
`device_registry.async_get(device_id)` to `vehicle_id`. I personally do not know.

File: custom_components/kia_uvo/__init__.py [hunk 4]
```diff
@@ -244,12 +262,29 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry):
...
+        ]
         config_update_listener()
 
-        hass.data[DOMAIN] = None
+        hass.data[DOMAIN][vehicle_identifier] = None
 
     return unload_ok
+
+
+async def async_migrate_entry(hass, config_entry: ConfigEntry):
```
Comment (fuatakgun) [inline custom_components/kia_uvo/__init__.py:279]: very nice approach, thanks for saving existing users

File: custom_components/kia_uvo/config_flow.py [hunk 5]
```diff
@@ -52,7 +44,7 @@
 
 
 class KiaUvoOptionFlowHandler(config_entries.OptionsFlow):
-    def __init__(self, config_entry):
+    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
```
Comment (fuatakgun) [inline custom_components/kia_uvo/config_flow.py:47]: 👏

File: custom_components/kia_uvo/config_flow.py [hunk 6]
```diff
@@ -177,3 +181,38 @@ async def async_step_user(self, user_input=None):
         return self.async_show_form(
             step_id="user", data_schema=self.schema, errors=errors
         )
+
+    async def async_step_pick_vehicle(
```
Comment (fuatakgun) [inline custom_components/kia_uvo/config_flow.py:185]: are we calling this function?

File: custom_components/kia_uvo/binary_sensor.py [hunk 7]
```diff
@@ -12,14 +12,20 @@
...
+    CONF_VEHICLE_IDENTIFIER,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
-    vehicle: Vehicle = hass.data[DOMAIN][DATA_VEHICLE_INSTANCE]
-
+    vehicle: Vehicle = hass.data[DOMAIN][config_entry.data[CONF_VEHICLE_IDENTIFIER]][
```
Comment (fuatakgun) [inline custom_components/kia_uvo/binary_sensor.py:26]: we can add a new field to `KiaUvoEntity.py` class as self.vehicle and all inherited classes will access to `self.vehicle` and we do not need to find correct vehicle under all entities one by one.

File: custom_components/kia_uvo/device_tracker.py [hunk 8]
```diff
@@ -7,13 +7,14 @@
...
 from .KiaUvoEntity import KiaUvoEntity
-from .const import DOMAIN, DATA_VEHICLE_INSTANCE
+from .const import DOMAIN, DATA_VEHICLE_INSTANCE, CONF_VEHICLE_IDENTIFIER
 
 _LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
-    vehicle: Vehicle = hass.data[DOMAIN][DATA_VEHICLE_INSTANCE]
+    vehicle_identifier = config_entry.data[CONF_VEHICLE_IDENTIFIER]
```
Comment (fuatakgun) [inline custom_components/kia_uvo/device_tracker.py:16]: repeat: we can add a new field to `KiaUvoEntity.py` class as self.vehicle and all inherited classes will access to `self.vehicle` and we do not need to find correct vehicle under all entities one by one.

File: custom_components/kia_uvo/lock.py [hunk 9]
```diff
@@ -4,13 +4,19 @@
...
+    VEHICLE_LOCK_ACTION,
+    CONF_VEHICLE_IDENTIFIER,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
-    vehicle: Vehicle = hass.data[DOMAIN][DATA_VEHICLE_INSTANCE]
+    vehicle_identifier = config_entry.data[CONF_VEHICLE_IDENTIFIER]
```
Comment (fuatakgun) [inline custom_components/kia_uvo/lock.py:18]: repeat: we can add a new field to `KiaUvoEntity.py` class as self.vehicle and all inherited classes will access to `self.vehicle` and we do not need to find correct vehicle under all entities one by one.

File: custom_components/kia_uvo/sensor.py [hunk 10]
```diff
@@ -23,13 +23,15 @@
...
     DYNAMIC_DISTANCE_UNIT,
+    CONF_VEHICLE_IDENTIFIER,
 )
 
 _LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
-    vehicle: Vehicle = hass.data[DOMAIN][DATA_VEHICLE_INSTANCE]
+    vehicle_identifier = config_entry.data[CONF_VEHICLE_IDENTIFIER]
```
Comment (fuatakgun) [inline custom_components/kia_uvo/sensor.py:33]: repeat: we can add a new field to `KiaUvoEntity.py` class as self.vehicle and all inherited classes will access to `self.vehicle` and we do not need to find correct vehicle under all entities one by one.

File: custom_components/kia_uvo/services.yaml [hunk 11]
```diff
@@ -1,10 +1,36 @@
...
+      name: Device
+      description: Device id of the vehicle.
+      required: false
+      selector:
+        device:
+          integration: kia_uvo
 start_climate:
-  description: TEST - Please use cautiously - start car and climate. Not all items available for all cars.  Refer to Hyundai or Kia app for which are supported for your car. 
+  description: >-
+    TEST - Please use cautiously - start car and climate. Not all items
```
Comment (fuatakgun) [inline custom_components/kia_uvo/services.yaml:23]: let's remove `TEST`, it is well established now, thanks to you

File: custom_components/kia_uvo/testAPI.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
+import logging
```
Comment (fuatakgun) [inline custom_components/kia_uvo/testAPI.py:1]: do you want to keep this file here?

**Review by fuatakgun — CHANGES_REQUESTED** [review]
overall looks good, we can squash some more improvements here.

File: custom_components/kia_uvo/config_flow.py (hunk 6 above)
Comment (cdnninja) [reply custom_components/kia_uvo/config_flow.py:185]: Not at this point.  That would be used once we have a get_vehicle in all APIs.  Once that is done can select the device and return that in the token.

File: custom_components/kia_uvo/testAPI.py (hunk 12 above)
Comment (cdnninja) [reply custom_components/kia_uvo/testAPI.py:1]: Ideally yes.  It is very helpful for testing this new feature so thought I would save the effort from others.

**Comment by cdnninja** [comment]
I will do a round and tweaks.  After running this for a few days I am seeing the occasional error around non unique keys on sensors.   I will have to dig into that. Restarting seems to resolve it.

**Comment by cdnninja** [comment]
For some reason it won't let me reply to the threads anymore.  In response to vehicle id being the same as device id.  Sort of.   Home assistants "device ID" is the car.  Which is our vehicle id.  I think device ID in the token is different and more of a "client id" representing the endpoint from an API standpoint.

**Pull request closed without merging**